cores/esp8266/core_esp8266_main.cpp

```diff
@@ -188,31 +193,25 @@ void init_done() {
 
 */
 
-#ifdef NO_EXTRA_4K_HEAP
-/* this is the default NONOS-SDK user's heap location */
-cont_t g_cont __attribute__ ((aligned (16)));
-#endif
-
-extern "C" void ICACHE_RAM_ATTR app_entry(void)
+extern "C" void ICACHE_RAM_ATTR app_entry_redefinable(void) __attribute__((weak));
+extern "C" void ICACHE_RAM_ATTR app_entry_redefinable(void)
 {
-#ifdef NO_EXTRA_4K_HEAP
-
-    /* this is the default NONOS-SDK user's heap location */
-    g_pcont = &g_cont;
-
-#else
-
     /* Allocate continuation context on this SYS stack,
        and save pointer to it. */
     cont_t s_cont __attribute__((aligned(16)));
     g_pcont = &s_cont;
 
-#endif
-
     /* Call the entry point of the SDK code. */
     call_user_start();
 }
 
+static void ICACHE_RAM_ATTR app_entry_custom (void) __attribute__((weakref("app_entry_redefinable")));
+
+extern "C" void ICACHE_RAM_ATTR app_entry (void)
```

Comment:
Can we drop this extra call level?

In this file, you define:

```c++
extern "C" void ICACHE_RAM_ATTR app_entry_sysstack(void)
{
    cont_t s_cont __attribute__((aligned(16)));
    g_pcont = &s_cont;
    call_user_start();
}
extern "C" void app_entry(void) __attribute__((weak, alias("app_entry_sysstack")));
```

In the other file, you define app_entry as a strong symbol:

```c++
extern "C" void ICACHE_RAM_ATTR app_entry(void)
{
    g_pcont = &g_cont;
    call_user_start();
}
```

I think it should work...

Reply:
I tried that, and just re-tried your way (weak+alias instead of weakref) and it still does not work, the strong one is always used. I don't know why for sure.

I had put that comment just above about that.
> (note: setting app_entry() itself as "weak" is not sufficient and always ends up with the other "noextra4k" one linked, maybe because it has a default value in linker scripts).

Reply:
Yes, you are right. That's a side effect of specifying something as ENTRY, it also makes the symbol undefined, and the linker picks it up from core_esp8266_app_entry_noextra4k.cpp before it sees core_esp8266_main.cpp (probably due to alphabetical ordering when creating the archive file)...

Reply:
I tried again (after dot_a_linkage, just to be sure), still the same. The indirection is still needed.

Reply:
No problem, that only cost us 16 bytes of the stack, which we can later remove by re-writing `app_entry` as a one-line assembly function which call0-s `app_entry_redefinable`.